### PR TITLE
[2871] Thread mode for MessageComposer

### PR DESCRIPTION
**Pull request by bychkovdmitry**
### 🎯 Goal

Add support for thread replies for the new `MessageComposer`

### 🎨 UI Changes

| Before | After |
| --- | --- |
| photo_2022-01-26 18 08 52 | photo_2022-01-26 18 05 51|
| photo_2022-01-26 18 09 03 |  photo_2022-01-26 18 05 49 |

#### General
- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] Assigned a person / code owner group (required)
- [x] Thread with the PR link started in a respective Slack channel (#android-chat-core or #android-chat-ui) (required)
- [ ] PR targets the `develop` branch
- [x] PR is linked to the GitHub issue it resolves

#### Code & documentation
- [ ] Changelog is updated with client-facing changes
- [ ] New code is covered by unit tests
- [x] Comparison screenshots added for visual changes
- [ ] Affected documentation updated (KDocs, docusaurus, tutorial)

### ☑️Reviewer Checklist
- [ ] UI Components sample runs & works
- [ ] Compose sample runs & works
- [ ] UI Changes correct (before & after images)
- [ ] Bugs validated (bugfixes)
- [ ] New feature tested and works
- [ ] Release notes and docs clearly describe changes
- [ ] All code we touched has new or updated KDocs

### 🎉 GIF

_Please provide a suitable gif that describes your work on this pull request_

### CODE REVIEW

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/composer/MessageComposerDefaultCenterContent.kt [hunk 1]
```diff
@@ -81,11 +81,9 @@ public class MessageComposerDefaultCenterContent : FrameLayout, MessageComposerC
      */
     private fun init() {
         binding = StreamUiMessageComposerDefaultCenterContentBinding.inflate(streamThemeInflater, this)
-        binding.messageEditText.addTextChangedListener(object : TextWatcherAdapter() {
-            override fun afterTextChanged(s: Editable) {
-                onTextChanged(s.toString())
-            }
-        })
+        binding.messageEditText.doAfterTextChanged { editable: Editable? ->
```
Comment (bychkovdmitry) [inline stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/composer/MessageComposerDefaultCenterContent.kt:84]: TextWatcherAdapter is internal here and produces a warning

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/composer/MessageComposerViewModel.kt [hunk 2]
```diff
@@ -1,22 +1,20 @@
...
 import io.getstream.chat.android.common.state.Reply
+import io.getstream.chat.android.common.state.ValidationError
 import kotlinx.coroutines.flow.Flow
 import kotlinx.coroutines.flow.MutableStateFlow
 import kotlinx.coroutines.flow.StateFlow
-import kotlinx.coroutines.flow.collect
-import kotlinx.coroutines.launch
 
 /**
  * ViewModel responsible for handling the composing and sending of messages.
```
Comment (bychkovdmitry) [inline stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/composer/MessageComposerViewModel.kt:20]: These ViewModels are almost the same now

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/composer/MessageComposerDefaultCenterContent.kt (hunk 1 above)
Comment (filbabic) [reply stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/composer/MessageComposerDefaultCenterContent.kt:84]: Do we lose out on any functionality without it?

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/composer/MessageComposerView.kt [hunk 3]
```diff
@@ -387,5 +414,5 @@ public class MessageComposerView : ConstraintLayout {
  * The [renderState] hook function is invoked when the state changes. You should update the UI of the view inside this function implementation.
  */
 public interface MessageComposerChild {
-    public fun renderState(state: MessageInputState)
+    public fun renderState(state: MessageComposerState)
```
Comment (filbabic) [inline stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/composer/MessageComposerView.kt:417]: Maybe add a small comment about hwat the `state` is? WDYT?

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/composer/MessageComposerViewModel.kt (hunk 2 above)
Comment (filbabic) [reply stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/composer/MessageComposerViewModel.kt:20]: Ideally, we'd just use one - I'm a bit skeptical as I feel we might come to a situation where there are more differences in the state/behavior due to XML requiring a different approach from Compose

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/composer/MessageComposerViewModel.kt [hunk 4]
```diff
@@ -32,61 +30,66 @@ public class MessageComposerViewModel(
...
+    public fun setMessageInput(value: String): Unit = messageComposerController.setMessageInput(value)
+
+    /**
+     * Called when the "Also send as a direct message" checkbox is checked or unchecked.
+     *
+     * @param alsoSendToChannel If the message will be shown in the channel after it is sent.
+     */
+    public fun setAlsoSendToChannel(alsoSendToChannel: Boolean): Unit =
+        messageComposerController.setAlsoSendToChannel(alsoSendToChannel)
+
```
Comment (filbabic) [inline stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/composer/MessageComposerViewModel.kt:92]: extra line

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/composer/MessageComposerViewModel.kt [hunk 5]
```diff
@@ -185,20 +173,20 @@ public class MessageComposerViewModel(
...
-        messageComposerController.selectCommand(command)
-        updateMessageInputState()
-    }
+    public fun selectCommand(command: Command): Unit = messageComposerController.selectCommand(command)
+
+    /**
+     * Toggles the visibility of the command suggestion list popup.
+     */
+    public fun toggleCommandsVisibility(): Unit = messageComposerController.toggleCommandsVisibility()
+
```
Comment (filbabic) [inline stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/composer/MessageComposerViewModel.kt:189]: here too

File: stream-chat-android-ui-components/src/main/res/layout/stream_ui_message_composer.xml [hunk 6]
```diff
@@ -35,10 +36,20 @@
...
             />
 
     </androidx.appcompat.widget.LinearLayoutCompat>
+
+    <FrameLayout
+        android:id="@+id/bottomContent"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        app:layout_constraintBottom_toBottomOf="parent"
+        app:layout_constraintLeft_toLeftOf="parent"
```
Comment (filbabic) [inline stream-chat-android-ui-components/src/main/res/layout/stream_ui_message_composer.xml:50]: we should use start/end for RTL

**Review by filbabic — COMMENTED** [review]
Left a comment or two but mostly this looks great!

Could we update the changelog and resolve the conversations? Once it's done, we can proceed with the merge!

### ☑️Reviewer Checklist
- [x] UI Components sample runs & works
- [x] Compose sample runs & works
- [x] UI Changes correct (before & after images)
- [ ] Bugs validated (bugfixes)
- [x] New feature tested and works
- [ ] Release notes and docs clearly describe changes
- [x] All code we touched has new or updated KDocs

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/composer/MessageComposerDefaultCenterContent.kt (hunk 1 above)
Comment (bychkovdmitry) [reply stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/composer/MessageComposerDefaultCenterContent.kt:84]: `TextWatcherAdapter` is just a noop adapter:

```
@RestrictTo(LIBRARY_GROUP)
public class TextWatcherAdapter implements TextWatcher {

  @Override
  public void beforeTextChanged(@NonNull CharSequence s, int start, int count, int after) {}

  @Override
  public void onTextChanged(@NonNull CharSequence s, int start, int before, int count) {}

  @Override
  public void afterTextChanged(@NonNull Editable s) {}
}

```

**Pull request merged**